From 9a7e32a5eb4180c6d66bdd4094e1bc62311a73ca Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Mon, 18 Aug 2014 19:58:38 -0400
Subject: [block] nvme: Configure support for block flush

Message-id: <1408391935-24886-6-git-send-email-dmilburn@redhat.com>
Patchwork-id: 87879
O-Subject: [RHEL7.1 PATCH BZ 1111259 05/22] NVMe: Configure support for block flush
Bugzilla: 1111259
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Keith Busch <keith.busch@intel.com>

This configures an nvme request_queue as flush capable if the device
has a volatile write cache present.

Signed-off-by: Keith Busch <keith.busch@intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
(cherry picked from commit a7d2ce2832d84e0182585f63bf96ca7323b3aee7)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/nvme-core.c b/drivers/block/nvme-core.c
index b6211d2..88d859d 100644
--- a/drivers/block/nvme-core.c
+++ b/drivers/block/nvme-core.c
@@ -1982,6 +1982,8 @@ static struct nvme_ns *nvme_alloc_ns(struct nvme_dev *dev, unsigned nsid,
  blk_queue_logical_block_size(ns->queue, 1 << ns->lba_shift);
  if (dev->max_hw_sectors)
   blk_queue_max_hw_sectors(ns->queue, dev->max_hw_sectors);
+ if (dev->vwc & NVME_CTRL_VWC_PRESENT)
+  blk_queue_flush(ns->queue, REQ_FLUSH | REQ_FUA);
 
  disk->major = nvme_major;
  disk->first_minor = 0;
@@ -2301,6 +2303,7 @@ static int nvme_dev_add(struct nvme_dev *dev)
  nn = le32_to_cpup(&ctrl->nn);
  dev->oncs = le16_to_cpup(&ctrl->oncs);
  dev->abort_limit = ctrl->acl + 1;
+ dev->vwc = ctrl->vwc;
  memcpy(dev->serial, ctrl->sn, sizeof(ctrl->sn));
  memcpy(dev->model, ctrl->mn, sizeof(ctrl->mn));
  memcpy(dev->firmware_rev, ctrl->fr, sizeof(ctrl->fr));
diff --git a/include/linux/nvme.h b/include/linux/nvme.h
index 50f1bc2..aa135f7 100644
--- a/include/linux/nvme.h
+++ b/include/linux/nvme.h
@@ -98,6 +98,7 @@ struct nvme_dev {
  u32 stripe_size;
  u16 oncs;
  u16 abort_limit;
+ u8 vwc;
  u8 initialized;
 };
 
diff --git a/include/uapi/linux/nvme.h b/include/uapi/linux/nvme.h
index faaedc3..bc64c9e 100644
--- a/include/uapi/linux/nvme.h
+++ b/include/uapi/linux/nvme.h
@@ -73,6 +73,7 @@ enum {
  NVME_CTRL_ONCS_COMPARE   = 1 << 0,
  NVME_CTRL_ONCS_WRITE_UNCORRECTABLE = 1 << 1,
  NVME_CTRL_ONCS_DSM   = 1 << 2,
+ NVME_CTRL_VWC_PRESENT   = 1 << 0,
 };
 
 struct nvme_lbaf {
-- 
1.7.1